js/ui/status/missionGame.js

```diff
@@ -0,0 +1,648 @@
+        this._textIndex = 0;
+        this._scrollTimer = 0;
+        this._waitCount = 0;
+        this.complete = false;
+
+        this._label.clutter_text.x_expand = true;
+        this.bubble.add(this._label, { x_fill: true, expand: true });
+        this.view.add(this.bubble, { x_fill: true, expand: true });
+    },
+    start: function(scrollView) {
+        /* Immediately display the first character so that
+         * scrolling to the bottom will work */
+        this._textIndex = 0;
+        this._textIndex++;
+        this._label.set_text(this._text.slice(0, this._textIndex));
+
+        /* Add a timeout to gradually display all this text */
+        this._startScrollAnimation(scrollView);
+    },
+    _startScrollAnimation: function(done) {
+        if (!this._scrollTimer) {
+            this._waitCount = 0;
+            this._scrollTimer = GLib.timeout_add(GLib.PRIORITY_DEFAULT, 10, Lang.bind(this, function() {
+                if (this._waitCount > 0) {
+                    this._waitCount--;
+                    return true;
+                }
+
+                if (this._textIndex === this._text.length) {
+                    this.complete = true;
+                    this.emit('finished-scrolled');
+                    this._scrollTimer = 0;
+                    return false;
+                }
+
+                this._textIndex++;
+                this._label.set_text(this._text.slice(0, this._textIndex));
+
+                /* Stop on punctuation */
+                if ('!?.'.indexOf(this._text[this._textIndex - 1]) !== -1) {
+                    this._waitCount = 100;
+                    return true;
+                }
+
+                /* Running this on a timer every time is not
+                 * ideal, but I haven't found a better way
+                 * ensure the view is reliably always
+                 * scrolled */
+                if (done) {
+                    done();
+                }
+
+                return true;
+            }));
+        }
+
+        return this._scrollTimer;
+    },
+    fastForward: function() {
+        this._textIndex = this._text.length - 1;
+    },
+    acceptAdditionalContent: function(content) {
+        if (content.text && content.text.length && content.type === 'scrolled') {
+            this._text = wrapTextWith(this._text + '\n' + content.text, WRAP_CONSTANT, '').join('\n');
+            this._startScrollAnimation();
+            return true;
+        }
+
+        return false;
+    }
+});
+Signals.addSignalMethods(ScrolledLabel.prototype);
+
+const WrappedLabel = new Lang.Class({
+    Name: 'WrappedLabel',
+    Extends: ScrolledLabel,
+    _init: function(params, settings) {
+        this.parent(params, settings);
+        this._text = wrapTextWith(settings.text, WRAP_CONSTANT, '> ').join("\n");
+        this.fastForward();
+    },
+    acceptAdditionalContent: function(content) {
+        if (content.text && content.text.length && content.type === 'wrapped') {
```

Comment:
I think `content.text` will evaluate to true if the length is not zero, so the second check is redundant